agiverse/agent/messaging.py

```diff
@@ -258,11 +258,72 @@ async def _process_message(self, message):
+        
+        if recent_steps:
+            working_memory_content = "\n".join(step.to_string() for step in recent_steps)
+        else:
+            working_memory_content = ""
         all_memories = await self.agent.memory_manager.memory_stream.get_all_memories()
         long_term_memory = await self._filter_and_rank_memories(working_memory_content, all_memories)
+        
+        return working_memory_content, [long_term_memory]
+
+    async def _get_chat_context(self, sender_id: str, limit: int = 5) -> List[Memory]:
+        chat_memories = await self.agent.memory_manager.memory_stream.get_memories_by_type('chat')
+        relevant_memories = [
+            memory for memory in chat_memories
+            if (str(sender_id) in memory.metadata['sender_id'])
+        ]
+        
+        return sorted(relevant_memories, key=lambda x: x.created_at, reverse=True)[:limit]
+
+    async def _get_system_context(self, limit: int = 5) -> List[Memory]:
+        system_memories = await self.agent.memory_manager.memory_stream.get_memories_by_type('system')
+        return sorted(system_memories, key=lambda x: x.created_at, reverse=True)[:limit]
+
+    async def _merge_memories_to_string(self, existing_memory: str, new_memories: List[Memory]) -> str:
+        if not isinstance(existing_memory, str):
+            existing_memory = ""
+            
+        memory_contents = [memory.content for memory in new_memories]
+        return existing_memory + "\n" + "\n".join(memory_contents) if existing_memory else "\n".join(memory_contents)
+
+    async def _get_current_message(self, mock_message: Optional[Dict] = None) -> Optional[Dict]:
+        try:
+            current_message = await self.message_queue.get_nowait()
+        except asyncio.QueueEmpty:
+            logger.warning("No message in queue when processing message")
+            current_message = None
+        
+        return mock_message or current_message
+
+    async def _process_chat_message(self, long_term_memory: List[str], current_message: Dict) -> List[str]:
+        sender_id = current_message.get('data', {}).get('senderID')
+        if sender_id:
+            chat_context = await self._get_chat_context(sender_id)
+            merged_memory = await self._merge_memories_to_string(long_term_memory[0], chat_context)
+            return [merged_memory]
+        return long_term_memory
+
+    async def _process_system_message(self, long_term_memory: List[str], current_message: Dict) -> List[str]:
+        channel_id = current_message.get('data', {}).get('channelID')
+        if channel_id:
+            system_context = await self._get_system_context(channel_id)
+            merged_memory = await self._merge_memories_to_string(long_term_memory[0], system_context)
+            return [merged_memory]
+        return long_term_memory
+
+    async def _generate_model_response(self, message: Optional[Dict] = None):
```

Comment:
Using last message is not enough because there could be (and often are) multiple messages between two model calls.
What would work is checking all self.messages (for chat message) and self.system_messages (for system message), and should handle the case where there are multiple messages and system_messages